### ISSUE TITLE
DNS to a Pod

### ISSUE THREAD

**Issue by stevesloka** [kubernetes/kubernetes#13552]
I'm looking for a way to TLS everything inside my cluster (even between pods). I have a need for a monitoring tool to reach out to each pod (not through a service) over TLS. My problem is allowing the cert to request correctly. I know SkyDNS only creates entries for svc's, would I have to extend the SkyDNS add-on to also create records for pods? Any other creative thoughts?

**Comment by thockin** [comment]
More from IRC:

>  I don't want a service really, just when a pod calls out to the master pod, it will say, I'm "pod-xyz", then the master will ask hey "pod-xyz" are you ok?

#260 covers a lot of issues around DNS for pods
#4825 focuses on hostname issues

@bgrant0607 and @smarterclayton since this is the crew that DNS keeps coming back to. :)

Those issues aside, I have already indicated that I'm not against DNS for pods.  It just has to be done in a way that doesn't preclude things like nominal services.  If we can do DNS in a way that doesn't involve hostname, maybe it's not so bad.  So the question is how would your pods know what to say in the registration.

We could:

a) create DNS for $pod_uid.$pod_namespace.pod.cluster.local - you can't get the pod UID from downward API today, but maybe we could expand that.  Do NOT set the hostname.

b) create DNS for $sanitize($pod_name).$pod_namespace.pod.cluster.local - you can get the pod name from downward API today, but you would have to sanitize it, since it can include dots (and adding dots will break search-path-relative lookups).  The sanitization would be a source of bugs unless we made a sanitize name available on downward API.  This would also encourage abuse since the DNS name would be knowable a priori.  Do NOT set the hostname.

c) create a new pod field just for DNS name and expose that on the downward API.  At that point, it's sort of no different than the pod UID, but it could be more human-friendly.  Do NOT set the hostname.

Then, do we need PTR records?

**Comment by smarterclayton** [comment]
Can we just give the pod a DNS name equal it its podIP?  Breaks when we
have long DNS names and IPv6.

I really don't want to have reverse lookup for every pod in the system, but
since we say pods are like mini vms and all the cloud providers give you a
DNS record for your VM...

I will say that pods as vms can be a dangerous mindset - we kind of want
pods to be process groups with an ip.  I sort of want to fight for no
identity outside of a service (only way you get a DNS name for a pod is
through a particular service) but that's mostly based on a gut reaction to
treating pods as something special outside of their involvement in a
service.  I am definitely leaning towards if something is important by
itself it should still be in a service of one, and pods without services
are created solely for higher level concepts to use.

On Sep 3, 2015, at 12:36 PM, Tim Hockin <notifications@github.com> wrote:

More from IRC:

I don't want a service really, just when a pod calls out to the master pod,
it will say, I'm "pod-xyz", then the master will ask hey "pod-xyz" are you
ok?

#260 <https://github.com/kubernetes/kubernetes/issues/260> covers a lot of
issues around DNS for pods
#4825 <https://github.com/kubernetes/kubernetes/issues/4825> focuses on
hostname issues

@bgrant0607 <https://github.com/bgrant0607> and @smarterclayton
<https://github.com/smarterclayton> since this is the crew that DNS keeps
coming back to. :)

Those issues aside, I have already indicated that I'm not against DNS for
pods. It just has to be done in a way that doesn't preclude things like
nominal services. If we can do DNS in a way that doesn't involve hostname,
maybe it's not so bad. So the question is how would your pods know what to
say in the registration.

We could:

a) create DNS for $pod_uid.$pod_namespace.pod.cluster.local - you can't get
the pod UID from downward API today, but maybe we could expand that. Do NOT
set the hostname.

b) create DNS for $sanitize($pod_name).$pod_namespace.pod.cluster.local -
you can get the pod name from downward API today, but you would have to
sanitize it, since it can include dots (and adding dots will break
search-path-relative lookups). The sanitization would be a source of bugs
unless we made a sanitize name available on downward API. This would also
encourage abuse since the DNS name would be knowable a priori. Do NOT set
the hostname.

c) create a new pod field just for DNS name and expose that on the downward
API. At that point, it's sort of no different than the pod UID, but it
could be more human-friendly. Do NOT set the hostname.

Then, do we need PTR records?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/13552#issuecomment-137505226>
.

**Comment by stevesloka** [comment]
Today I'm using IP's to talk to each pod, but I have a need to enable TLS over that. When I gen my cert, if I need a dns name that I can add to my SAN. 

I don't care what the name of the pod is or how it's gen'd, just as long as I can use something like: <pod>.<namespace>.pod.k8s.local and it routes to that pod, then I can add *.<namespace>.pod.k8s.local to the cert's SAN. 

I also don't need for every pod, only for those that I maybe give a label to or something. 

I'm going to play with the Kube2DNS and see if I can mock up something. I could look at deploying another service discovery tool like Consul or something, but that is last ditch.

**Comment by thockin** [comment]
I spoke with Brian on this and we agreed on a reasonable compromise.  It
seems that Clayton came to the same solution.

For every pod, assign a DNS A record (PTR is different can of worms) of the
form `12-34-56-78.your-namespace.pod.cluster.local`.  The pod IP is already
part of the downward API and easily discovered besides.  This does not
expose any information that was not already knowable given the IP except
maybe the existence of the pod at all.  Search paths will allow you to
resolve "12-34-56-78.your-namespace.pod".  Note that the IP and
"your-namespace" come from the pod spec but "pod" is a literal and
"cluster.local" is the cluster domain suffix.

This sounds like it will satisfy your needs - please confirm?  This should
not be very hard to implement and if you have time to do it, you can
scratch your own itch much sooner than we can :)

Tim

**Comment by stevesloka** [comment]
Yup that sounds great! I'm happy to work on this. I will get some things together and maybe ask for a code review since I'm new to contributing back code-wise.

**Comment by bgrant0607** [comment]
I'm fine with the compromise, but am curious why a headless service won't work.
https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/services.md#headless-services

**Comment by bgrant0607** [comment]
The other thing we've agreed we should do is something along the lines of:
https://github.com/kubernetes/kubernetes/issues/260#issuecomment-133318903

I'm in agreement philosophically with: https://github.com/kubernetes/kubernetes/issues/13552#issuecomment-137537731

**Comment by stevesloka** [comment]
@bgrant0607 headless won't work since I need a way to talk to each pod directly for a sort of health check / stats update. I want to do that over TLS, so need a cert which can wildcard the dns. If I drew up a diagram or something, would that help?

**Comment by thockin** [comment]
From the IRC conversation: he wants to to every single pod.

**Comment by bgrant0607** [comment]
Headless generates a DNS record for every endpoint:
https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/kube2sky/kube2sky.go#L162

A single headless service could match every pod in a namespace.

**Comment by thockin** [comment]
Headless generates an A record for the set of endpoints, but that's not
useful if you are trying to connect to a particular instance to gather
metrics.  It also means that every pod has to have some common label or
else not get scraped.

**Comment by bgrant0607** [comment]
Ah. I thought we were going to generate As for the individuals, also.

A common label doesn't seem like a huge burden.

Anyway, I'm fine with the compromise, but was trying to understand why it was needed.

Thanks.

**Comment by streamnsight** [comment]
I'm deploying Zookeeper and Kafka on Kubernetes and the DNS / hostname issue is a nightmare:
Zookeeper takes other nodes' IPs, but services calling it most often are designed to receive a list of zk hosts (for failover). 
Kafka nodes define their own hostname in the config, and nodes talk to each other using hostnames. Services using Kafka also get passed the hostname... 
So there are 2 problems here: resolving the hostname (which seems to resolve to pod id anyways) but also keeping up-to-date. 
In my case so far I have had to update the /etc/hosts file in zk, kafka and all services depending on those to be able to use them properly, but keeping them up-to-date is another story. 

another issue I encountered is that zk and kafka data is persisted to disk (I mount an emptyDir) but if their IDs change, the data doesn't, and I end up with data that refer to nodes that are no longer, so i resorted to using node IPs (without the dots) as an ID for zk and Kafka, so that when an instance restarts on the same node (but with a different pod IP) and picks up the old data from the emptyDir (since emptyDir actually persist after a restart), the ID associated with the data is consistent.

DNS for pods would really help keep pod references straight across these type of distributed services.

**Comment by stevesloka** [comment]
I have a couple random questions about the DNS service: 

1. Why is the refresh interval 30 minutes? (https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/kube2sky/kube2sky.go#L61) So if I create / update / or delete a service, it won't be reflected until the next cycle every 30 mins? 

2. If I want to scale the service, we'll rely on each instance having the same data stored in etcd? I guess if I didn't want that, just need to scale etcd and deal with that complexity.

**Comment by smarterclayton** [comment]
On Sep 4, 2015, at 9:06 AM, Steve Sloka <notifications@github.com> wrote:

I have a couple random questions about the DNS service:

   1.

   Why is the refresh interval 30 minutes? (
   https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/kube2sky/kube2sky.go#L61)
   So if I create / update / or delete a service, it won't be reflected until
   the next cycle every 30 mins?



Service IPs are expected not to change over time.  So it's expected that
updates to services don't change things that need to be refreshed.




   1.

   If I want to scale the service, we'll rely on each instance having the
   same data stored in etcd? I guess if I didn't want that, just need to scale
   etcd and deal with that complexity.



DNS is unlikely to be running in a pod for much longer.  We're probably
going to move it to the nodes as a function of the service proxy, because
the service proxies already have the same info.  There is a PR open for it
but it's blocked on the other service proxy refractors.  When we do that
etcd will no longer be used - skydns will be backed directly by the Kube
API.




—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/13552#issuecomment-137730482>
.

**Comment by streamnsight** [comment]
@smarterclayton 

RE: Zookeeper from fabrik8

They take a different and simpler approach that has its own limitations: they basically have to declare a RC and a SERVICE per ZK instance: 
If you want a standalone ZK, you declare one, you want 3, you declare 3 and spin them manually, but you need to declare IDs as ENV variables here, so it cannot scale automatically, and each client application needs to be aware of each ZK service for this to work. -> doesn't scale.

That may not be such a problem with ZK since one usually don't need to scale 'automatically', but it's more of a problem for Kafka: this approach is just not convenient at all, and that defeats the purpose of using Kubernetes. I can (and have) do this in Docker alone! It's moving the problem to the client. 

So, it's true trying to make it scalable is a whole lot more complicated, but this is where i think Per pod DNS would just solve the problem.

**Comment by smarterclayton** [comment]
Ok, that helps to know that this is about Kafka.  The nominal services work
applies to this problem then (possibly the volume allocation issue for the
subset of Kafka users that want to guarantee the volume isn't lost on node
loss).

On Sep 4, 2015, at 12:31 PM, streamnsight <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton>

RE: Zookeeper from fabrik8

They take a different and simpler approach that has its own limitations:
they basically have to declare a RC and a SERVICE per ZK instance:
If you want a standalone ZK, you declare one, you want 3, you declare 3 and
spin them manually, but you need to declare IDs as ENV variables here, so
it cannot scale automatically, and each client application needs to be
aware of each ZK service for this to work. -> doesn't scale.

That may not be such a problem with ZK since one usually don't need to
scale 'automatically', but it's more of a problem for Kafka: this approach
is just not convenient at all, and that defeats the purpose of using
Kubernetes. I can (and have) do this in Docker alone! It's moving the
problem to the client.

So, it's true trying to make it scalable is a whole lot more complicated,
but this is where i think Per pod DNS would just solve the problem.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/13552#issuecomment-137783333>
.

**Comment by thockin** [comment]
@stevesloka 

> Why is the refresh interval 30 minutes? (https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/kube2sky/kube2sky.go#L61) So if I create / update / or delete a service, it won't be reflected until the next cycle every 30 mins?

The refresh is a forced refresh.  The watch triggers whenever there's an update in real time.  The forced refresh is a safety net that we don't really need.

> If I want to scale the service, we'll rely on each instance having the same data stored in etcd? I guess if I didn't want that, just need to scale etcd and deal with that complexity.

We treat etcd as an implementation detail.  If you want to scale the DNS service, just add more replicas - they are fungible.

**Comment by stevesloka** [comment]
@thockin thanks for the clarification!

**Comment by smarterclayton** [comment]
Steve, for your kafka use case are you planning on using persistent volumes
for the kafka broker pods or not?

On Tue, Sep 8, 2015 at 11:08 AM, Steve Sloka <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> thanks for the clarification!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/issues/13552#issuecomment-138595070>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by stevesloka** [comment]
@smarterclayton I wasn't working on kafka, but @streamnsight was, I'm looking to integrate to the SkyDNS pieces currently.

**Comment by smarterclayton** [comment]
I bring that up because we think in order to make persistent volumes work
for nominal services (where you can reuse the same volume when your pod
moves) your volume (if your app records hostname) would have the old
hostname (pod-1-1-1-1) not the new pod hostname, and the two wouldn't
match.  But I suspect most people aren't trying to use persistent volumes
yet.

On Tue, Sep 8, 2015 at 11:37 AM, Steve Sloka <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> I wasn't working on
> kafka, but @streamnsight <https://github.com/streamnsight> was, I'm
> looking to integrate to the SkyDNS pieces currently.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/issues/13552#issuecomment-138602974>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by streamnsight** [comment]
Just a quick update here:
Zookeeper seems to be the only one that really needs to resolve hostnames. I am looking to figure how to just use IPs and not hostnames.
For Kafka, I figured a way to define the advertised name as an IP and it seems to work OK but i need to do more testing.
Originally i had to fill in the hosts file to resolve hostnames, but apparently, by setting both
host.name 
and 
advertised.host.name
to the IP of the pod, kafka seems to be able to resolve its peers properly.
the issue remains with the kafka broker IDs which are associated with partitions, so when a pod restarts the data doesn't match the active brokers and one need to manually re balance partitions on the new broker.

**Comment by bgrant0607** [comment]
The Zookeeper use case belongs in #260 rather than this issue. It's a totally different scenario that will have a different solution.
